File: red/src/workflow.ts
// Lifecycle graph, preflight and backend advice, the port of
// io.github.getcolors.rybbit.workflow.

import { readPars, parName } from "red/cli";
import * as dryRun from "red/dry-run";
import { preflight } from "red/lifecycle";
import * as progress from "red/progress";
import * as tofu from "red/tofu";
import { adviceAdd, workflow, type Opts, type WireDecl } from "red/workflow";
import * as tools from "./tools.ts";
import * as validate from "./validate.ts";

export const defaults: Opts = {
  "provider-compute": "digitalocean", "provider-dns": "cloudflare",
  "provider-backend": "local", "compute-prevent-destroy": true,
  workdir: ".colors",
};

// Compute params recorded in the infrastructure state; undefined when the
// state holds none. An unreadable backend throws — the delete path treats that
// as fatal rather than falling back to the documentation address.
export async function stateOutput(opts: Opts): Promise<Record<string, unknown> | undefined> {
  const outputs = await tofu.outputs(
    tools.toolDir(opts, tools.infrastructureTool),
    tools.backendCredentialEnv(opts),
  );
  const params = (outputs as Record<string, unknown> | undefined)?.params;
  return params && typeof params === "object" ? params as Record<string, unknown> : undefined;
}

// A real delete runs the ansible cleanup before the infrastructure step, so
// the instance address must come out of the existing state here. An explicit
// `ip` (COLORS_PAR_IP) skips the read; a readable state without compute params
// leaves `ip` unset and the cleanup step skips itself; an unreadable backend
// fails loudly — swallowing it is how a live teardown ended up converging
// against 192.0.2.10.
export async function adoptState(
  opts: Opts,
  stateOutputFn: typeof stateOutput = stateOutput,
): Promise<Opts> {
  if (opts.ip) return { ...opts, "red/exit": 0 };
  try {
    return { ...opts, ...(await stateOutputFn(opts) ?? {}), "red/exit": 0 };
  } catch (err) {
    const message = err instanceof Error ? err.message : String(err);
    return {
      ...opts, "red/exit": 1,
      "red/err": "could not read the infrastructure state for " +
        `the delete cleanup: ${message}\n` +
        "fix the backend credentials, or supply " +
        `${parName("ip")} to address the instance directly`,
    };
  }
}

export async function startStep(
  opts: Opts,
  env: Record<string, string | undefined> = process.env,
  stateOutputFn: typeof stateOutput = stateOutput,
): Promise<Opts> {
  return preflight(opts, {
    defaults,
    overlay: readPars,
    validators: [
      (_opts, environment) => validate.envErrors(environment),
      (current) => validate.stateErrors(current),
      (current, _environment, { event, real }) =>
        real && (event === "create" || event === "delete")
          ? validate.secretErrors(current)
          : [],
      (current, _environment, { event, real }) =>
        real && event === "delete" && current["compute-prevent-destroy"]
          ? [`compute destruction is protected; set ${parName("compute-prevent-destroy")}=false to delete`]
          : [],
    ],
    afterValidate: async (current, _environment, { event, real }) =>
      real && event === "delete"
        ? adoptState(current, stateOutputFn)
        : { ...current, "red/exit": 0 },
  }, env);
}

export function wireFn(step: string, runOpts: Opts): WireDecl | undefined {
  if (runOpts["red/event"] === "delete") {
    const graph: Record<string, WireDecl> = {
      "rybbit/start": [startStep, "rybbit/ansible"],
      "rybbit/ansible": [tools.ansibleStep, "rybbit/dns"],
      "rybbit/dns": [tools.dnsStep, "rybbit/infrastructure"],
      "rybbit/infrastructure": [tools.infrastructureStep],
    };
    return graph[step];
  }
  const graph: Record<string, WireDecl> = {
    "rybbit/start": [startStep, "rybbit/infrastructure"],
    "rybbit/infrastructure": [tools.infrastructureStep, "rybbit/dns"],
    "rybbit/dns": [tools.dnsStep, "rybbit/ansible"],
    "rybbit/ansible": [tools.ansibleStep, "rybbit/acceptance"],
    "rybbit/acceptance": [tools.acceptanceStep],
  };
  return graph[step];
}

export function backendAdvice(tool: string) {
  return tofu.conventionalBackendAdvice({
    dir: (opts) => tools.toolDir(opts, tool),
    key: (opts) => `${opts.profile}/${tool}.tfstate`,
  });
}

export const sideEffecting = [
  "rybbit/infrastructure", "rybbit/dns", "rybbit/ansible", "rybbit/acceptance",
];

function create() {
  let wf = workflow({ start: "rybbit/start", wireFn });
  wf = adviceAdd(wf, "rybbit/infrastructure", "before",
    "io.github.getcolors.rybbit.workflow/backend",
    backendAdvice(tools.infrastructureTool));
  wf = adviceAdd(wf, "rybbit/dns", "before",
    "io.github.getcolors.rybbit.workflow/backend",
    backendAdvice(tools.dnsTool));
  return dryRun.advise(progress.advise(wf), sideEffecting);
}

export const rybbitWorkflow = create();

File: blue/src/package_rybbit_blue/workflow.py
"""The graph, the port of io.github.getcolors.rybbit.workflow."""

from __future__ import annotations

from blue import dry_run, progress, tofu
from blue.cli import par_name, read_pars
from blue.lifecycle import preflight
from blue.workflow import advice_add, workflow

from . import tools, validate

DEFAULTS = {"provider-compute": "digitalocean", "provider-dns": "cloudflare",
            "provider-backend": "local", "compute-prevent-destroy": True,
            "workdir": ".colors"}


async def state_output(opts: dict) -> dict | None:
    """Compute params recorded in the infrastructure state; None when the state
    holds none. An unreadable backend raises — the delete path treats that as
    fatal rather than falling back to the documentation address."""
    outputs = await tofu.outputs(tools.tool_dir(opts, tools.infrastructure_tool),
                                 tools.backend_credential_env(opts))
    return (outputs or {}).get("params")


async def adopt_state(opts: dict) -> dict:
    """A real delete runs the ansible cleanup before the infrastructure step,
    so the instance address must come out of the existing state here. An
    explicit `ip` (COLORS_PAR_IP) skips the read; a readable state without
    compute params leaves `ip` unset and the cleanup step skips itself; an
    unreadable backend fails loudly — swallowing it is how a live teardown
    ended up converging against 192.0.2.10."""
    if opts.get("ip"):
        return {**opts, "blue/exit": 0}
    try:
        return {**opts, **((await state_output(opts)) or {}), "blue/exit": 0}
    except Exception as e:
        return {**opts, "blue/exit": 1,
                "blue/err": ("could not read the infrastructure state for "
                             f"the delete cleanup: {e}\n"
                             "fix the backend credentials, or supply "
                             f"{par_name('ip')} to address the instance directly")}


async def start_step(original: dict, env: dict | None = None) -> dict:
    async def after(opts, _env, context):
        if context["real"] and context["event"] == "delete":
            return await adopt_state(opts)
        return {**opts, "blue/exit": 0}

    return await preflight(
        original, defaults=DEFAULTS, overlay=read_pars, env=env,
        validators=[
            lambda _o, e, _c: validate.env_errors(e),
            lambda o, _e, _c: validate.state_errors(o),
            lambda o, _e, c: (validate.secret_errors(o)
                              if c["real"] and c["event"] in ("create", "delete") else []),
            lambda o, _e, c: ([f"compute destruction is protected; set "
                               f"{par_name('compute-prevent-destroy')}=false to delete"]
                              if c["real"] and c["event"] == "delete"
                              and o.get("compute-prevent-destroy") else []),
        ],
        after_validate=after)


def wire_fn(step: str, run_opts: dict):
    if run_opts.get("blue/event") == "delete":
        return {
            "rybbit/start": (start_step, "rybbit/ansible"),
            "rybbit/ansible": (tools.ansible_step, "rybbit/dns"),
            "rybbit/dns": (tools.dns_step, "rybbit/infrastructure"),
            "rybbit/infrastructure": (tools.infrastructure_step,),
        }.get(step)
    return {
        "rybbit/start": (start_step, "rybbit/infrastructure"),
        "rybbit/infrastructure": (tools.infrastructure_step, "rybbit/dns"),
        "rybbit/dns": (tools.dns_step, "rybbit/ansible"),
        "rybbit/ansible": (tools.ansible_step, "rybbit/acceptance"),
        "rybbit/acceptance": (tools.acceptance_step,),
    }.get(step)


def backend_advice(tool: str):
    return tofu.conventional_backend_advice(
        dir=lambda o, tool=tool: tools.tool_dir(o, tool),
        key=lambda o, tool=tool: f"{o.get('profile')}/{tool}.tfstate")


side_effecting = ["rybbit/infrastructure", "rybbit/dns", "rybbit/ansible",
                  "rybbit/acceptance"]


def create_workflow():
    wf = workflow(start="rybbit/start", wire_fn=wire_fn)
    wf = advice_add(wf, "rybbit/infrastructure", "before",
                    "io.github.getcolors.rybbit.workflow/backend",
                    backend_advice(tools.infrastructure_tool))
    wf = advice_add(wf, "rybbit/dns", "before",
                    "io.github.getcolors.rybbit.workflow/backend",
                    backend_advice(tools.dns_tool))
    return dry_run.advise(progress.advise(wf), side_effecting)


rybbit_workflow = create_workflow()
